File: python-orders/tracing.py
from opentelemetry import trace
from opentelemetry import metrics

from opentelemetry.sdk.resources import Resource

from opentelemetry.sdk.trace import TracerProvider

from opentelemetry.sdk.trace.export import (
    BatchSpanProcessor
)

from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
    OTLPSpanExporter
)

from opentelemetry.sdk.metrics import (
    MeterProvider
)

from opentelemetry.sdk.metrics.export import (
    PeriodicExportingMetricReader
)

from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import (
    OTLPMetricExporter
)


resource = Resource.create(
    {
        "service.name": "python-orders"
    }
)

#
# Traces
#

provider = TracerProvider(
    resource=resource
)

processor = BatchSpanProcessor(
    OTLPSpanExporter(
        endpoint="otel-collector:4317",
        insecure=True
    )
)

provider.add_span_processor(
    processor
)

trace.set_tracer_provider(
    provider
)

#
# Metrics
#

metric_reader = (
    PeriodicExportingMetricReader(
        OTLPMetricExporter(
            endpoint="otel-collector:4317",
            insecure=True
        ),
        export_interval_millis=10000
    )
)

meter_provider = MeterProvider(
    resource=resource,
    metric_readers=[
        metric_reader
    ]
)

metrics.set_meter_provider(
    meter_provider
)

print(
    "OpenTelemetry initialized"
)

File: node-frontend/tracing.js
const {
  NodeSDK,
} = require(
  "@opentelemetry/sdk-node"
);

const {
  getNodeAutoInstrumentations,
} = require(
  "@opentelemetry/auto-instrumentations-node"
);

const {
  OTLPTraceExporter,
} = require(
  "@opentelemetry/exporter-trace-otlp-http"
);

const {
  OTLPMetricExporter,
} = require(
  "@opentelemetry/exporter-metrics-otlp-http"
);

const {
  OTLPLogExporter,
} = require(
  "@opentelemetry/exporter-logs-otlp-http"
);

const {
  BatchLogRecordProcessor,
} = require(
  "@opentelemetry/sdk-logs"
);

const {
  PeriodicExportingMetricReader,
} = require(
  "@opentelemetry/sdk-metrics"
);

const traceExporter =
  new OTLPTraceExporter({
    url:
      "http://otel-collector:4318/v1/traces",
  });

const metricExporter =
  new OTLPMetricExporter({
    url:
      "http://otel-collector:4318/v1/metrics",
  });

const logExporter =
  new OTLPLogExporter({
    url:
      "http://otel-collector:4318/v1/logs",
  });

const metricReader =
  new PeriodicExportingMetricReader({
    exporter:
      metricExporter,
    exportIntervalMillis:
      10000,
  });

const sdk =
  new NodeSDK({
    serviceName:
      "node-frontend",

    traceExporter,

    metricReader,

    logRecordProcessors: [
      new BatchLogRecordProcessor(
        logExporter
      ),
    ],

    instrumentations: [
      getNodeAutoInstrumentations(),
    ],
  });

sdk.start();

console.log(
  "OpenTelemetry initialized"
);